src/System.Linq/src/System/Linq/Concatenate.cs

```diff
@@ -84,20 +94,20 @@ internal ConcatNIterator(ConcatIterator<TSource> previousConcat, IEnumerable<TSo
 
             public override Iterator<TSource> Clone()
             {
-                return new ConcatNIterator<TSource>(_previousConcat, _next, _nextIndex);
+                return new ConcatNEnumerableIterator<TSource>(_previousConcat, _next, _nextIndex);
             }
 
             internal override ConcatIterator<TSource> Concat(IEnumerable<TSource> next)
             {
                 if (_nextIndex == int.MaxValue - 2)
                 {
-                    // In the unlikely case of this many concatenations, if we produced a ConcatNIterator
+                    // In the unlikely case of this many concatenations, if we produced a ConcatNEnumerableIterator
                     // with int.MaxValue then state would overflow before it matched it's index.
                     // So we use the naïve approach of just having a left and right sequence.
-                    return new Concat2Iterator<TSource>(this, next);
+                    return new Concat2EnumerableIterator<TSource>(this, next);
                 }
 
-                return new ConcatNIterator<TSource>(this, next, _nextIndex + 1);
+                return new ConcatNEnumerableIterator<TSource>(this, next, _nextIndex + 1);
             }
 
             internal override IEnumerable<TSource> GetEnumerable(int index)
```

Comment:
Can't believe this was never caught by the tests, but we also have to change this function to check for the new iterators as well.